s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/S3OffsetManagerEntry.java

```diff
@@ -0,0 +1,227 @@
+     */
+    private S3OffsetManagerEntry(final Map<String, Object> properties) {
+        data = new HashMap<>(properties);
+        for (final String field : RESTRICTED_KEYS) {
+            if (data.get(field) == null) {
+                throw new IllegalArgumentException("Missing '" + field + "' property");
+            }
+        }
+    }
+
+    /**
+     * Creates an S3OffsetManagerEntry. Will return {@code null} if properties is {@code null}.
+     *
+     * @param properties
+     *            the properties to wrap. May be {@code null}.
+     * @return an S3OffsetManagerEntry.
+     * @throws IllegalArgumentException
+     *             if one of the {@link #RESTRICTED_KEYS} is missing.
+     */
+    @Override
+    public S3OffsetManagerEntry fromProperties(final Map<String, Object> properties) {
+        return wrap(properties);
+    }
+
+    @Override
+    public Object getProperty(final String key) {
+        if (RECORD_COUNT.equals(key)) {
+            return recordCount;
+        }
+        return data.get(key);
+    }
+
+    @Override
+    public void setProperty(final String property, final Object value) {
+        if (RESTRICTED_KEYS.contains(property)) {
+            throw new IllegalArgumentException(
+                    String.format("'%s' is a restricted key and may not be set using setProperty()", property));
+        }
+        data.put(property, value);
+    }
+
+    @Override
+    public void incrementRecordCount() {
+        recordCount++;
+    }
+
+    /**
+     * Gets the umber of records extracted from data returned from S3.
+     *
+     * @return the umber of records extracted from data returned from S3.
+     */
+    public long getRecordCount() {
+        return recordCount;
+    }
+
+    /**
+     * Gets the S3Object key for the current object.
+     *
+     * @return the S3ObjectKey.
+     */
+    public String getKey() {
+        return (String) data.get(OBJECT_KEY);
+    }
+
+    @Override
+    public int getPartition() {
+        return getInt(PARTITION);
+    }
+
+    @Override
+    public String getTopic() {
+        return getString(TOPIC);
+    }
+
+    /**
+     * Gets the S3 bucket for the current object.
+     *
+     * @return the S3 Bucket for the current object.
+     */
+    public String getBucket() {
+        return getString(BUCKET);
+    }
+    /**
+     * Creates a new offset map. No defensive copy is necessary.
+     *
+     * @return a new map of properties and values.
+     */
+    @Override
+    public Map<String, Object> getProperties() {
+        final Map<String, Object> result = new HashMap<>(data);
+        result.put(RECORD_COUNT, recordCount);
+        return result;
+    }
+    /**
+     * Returns the OffsetManagerKey for this Entry.
+     *
+     * @return the OffsetManagerKey for this Entry.
+     */
+    @Override
+    public OffsetManager.OffsetManagerKey getManagerKey() {
+        return () -> Map.of(BUCKET, data.get(BUCKET), OBJECT_KEY, data.get(OBJECT_KEY));
```

Comment:
Instead of objectkey storing as keys, it is better to store partition ids in key.
We will have fewer number of keys.

Just verified lenses s3 source connector and adobe s3 source connector, and they store partitionids.

Can we think about this too ?

Reply:
topic.partitions we have this config. Our earlier implementation was based on this.

Reply:
@gharris1727 your suggestion will be helpful here. 
According to javadocs of OffsetStorageReader : offsets() method, I was thinking we would have to store topic and partition id in offset storage keys atleast ?
 
```
@Override
public OffsetManager.OffsetManagerKey getManagerKey() {
    return () -> Map.of(BUCKET, data.get(BUCKET), TOPIC, TOPIC, PARTITION, PARTITION);
}
```

When we have several objects under specified topics and partitions and to retrieve the stored offset map, how can be better structure the keys ?